### PR TITLE
Add retry for field caps node requests

### PR DESCRIPTION
**Pull request by dnhatn**
This change targets the feature branch: `group-field-caps` (based on 7.x).

This adds a retry mechanism for node-based field caps requests introduced in #77047. Merging index responses on data nodes will be implemented in a follow-up.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:342]: this should be always != null when shard.curentNodeId != null, correct?

**Review by ywelsch — COMMENTED** [review]
Thanks Nhat. I've done just a quick pass today (didn't get further). I'm wondering if some of the retry logic around shard selection / grouping can be unit-tested (e.g. we currently test that retries ARE happening, but don't test how many etc).

**Comment by dnhatn** [comment]
@ywelsch Thank you for your review. All good points - I am addressing them.

**Comment by dnhatn** [comment]
@ywelsch I think I have addressed your feedback. I will add some more tests to RequestDispatcher. Would you mind taking another look?

I think we can use the existing `FieldCapabilitiesRequest` instead of introducing `FieldCapabilitiesNodeRequest` when the merging response is implemented. I will consider this in the merge response PR.

**Comment by dnhatn** [comment]
run elasticsearch-ci/bwc

**Review by jtibshirani — COMMENTED** [review]
Thanks for tackling this @dnhatn. I like that we have a dedicated class now to handle the request dispatching logic. The test coverage also looks great.

The one part I wasn't sure of was the synchronization strategy in `RequestDispatcher`. There is quite a bit of logic guarded under synchronized blocks, especially the one in `execute`. I wonder if it'd be better (and if it's even possible) to rely on atomic integers/ thread-safe collections for this? I haven't identified a concrete concern, just raising it to hear your thoughts.

**Comment by dnhatn** [comment]
@ywelsch Please hold off on the review. I am working on merging the responses, and I will integrate it in this PR.

**Comment by dnhatn** [comment]
> I wonder if it'd be better (and if it's even possible) to rely on atomic integers/ thread-safe collections for this? I haven't identified a concrete concern, just raising it to hear your thoughts

@jtibshirani Yes, we can go without synchronization.

**Review by dnhatn — COMMENTED** [review]
@ywelsch I've updated this PR. The RequestDispatcher and its tests are ready. The merging logic is still WIP. I need to discuss it with you before completing it. Would you please review the RequestDispatcher and the approach of the merging results logic?  I will take a look at your can_match PR tomorrow. Sorry for the delay - I've been focusing on this PR. Thank you!

**Review by ywelsch — COMMENTED** [review]
> I am working on merging the responses, and I will integrate it in this PR.

Let's revert that part. It has become too difficult to review this PR, and I think we will need more discussions on the merging logic. Let's not block the node-level action on this, but create a clear list of follow-ups.

**Comment by dnhatn** [comment]
@ywelsch @jtibshirani Thanks for reviews. This is ready again after I removed the merging logic. Would you mind taking another look?

**Review by ywelsch — APPROVED** [review]
LGTM

**Review by jtibshirani — COMMENTED** [review]
This looks good to me too, I just left some small comments.

**Review by jtibshirani — APPROVED** [review]
Looks good to me 🎉

**Comment by dnhatn** [comment]
@ywelsch @jtibshirani Thanks so much for your reviews.

**Pull request merged**
